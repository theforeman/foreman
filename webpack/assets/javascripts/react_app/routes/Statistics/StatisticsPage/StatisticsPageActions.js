import API from '../../../redux/API/API';

import {
  STATISTICS_PAGE_DATA_RESOLVED,
  STATISTICS_PAGE_DATA_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_PAGE_URL,
} from '../constants';

export const getStatisticsMeta = (
  url = STATISTICS_PAGE_URL
) => async dispatch => {
  const onFetchSuccess = ({ data }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_DATA_RESOLVED,
      payload: {
        metadata: data.charts,
        hasData: data.charts.length > 0,
        discussionUrl: data.discussion_url,
      },
    });
  };

  const onFetchError = ({ message }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_DATA_FAILED,
      payload: {
        message: {
          type: 'error',
          text: message,
        },
      },
    });
  };
  try {
    const response = await API.get(url);
    return onFetchSuccess(response);
  } catch (error) {
    return onFetchError(error);
  }
};

const hideLoading = () => ({
  type: STATISTICS_PAGE_HIDE_LOADING,
});
