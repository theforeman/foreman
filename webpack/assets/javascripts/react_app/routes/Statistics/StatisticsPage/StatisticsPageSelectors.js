export const selectStatisticsPage = state => state.statisticsPage;

export const selectStatisticsMetadata = state =>
  selectStatisticsPage(state).metadata;
export const selectStatisticsDiscussionUrl = state =>
  selectStatisticsPage(state).discussionUrl;
export const selectStatisticsIsLoading = state =>
  selectStatisticsPage(state).isLoading;
export const selectStatisticsMessage = state =>
  selectStatisticsPage(state).message;
export const selectStatisticsHasError = state =>
  selectStatisticsPage(state).hasError;
export const selectStatisticsHasMetadata = state =>
  selectStatisticsPage(state).hasData;
