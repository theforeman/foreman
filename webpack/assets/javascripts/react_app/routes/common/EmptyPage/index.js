import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from '../../../common/I18n';
import DefaultEmptyState from '../../../components/common/EmptyState';
import './emptypage.scss';

const EmptyPage = ({ message: { type, text } }) => (
  <DefaultEmptyState
    icon={type === 'error' ? 'error-circle-o' : 'add-circle-o'}
    header={type === 'error' ? __('Error') : __('No Results')}
    description={text}
  />
);

EmptyPage.propTypes = {
  message: PropTypes.shape({
    type: PropTypes.oneOf(['empty', 'error']),
    text: PropTypes.string,
  }),
};

EmptyPage.defaultProps = {
  message: PropTypes.shape({
    type: 'empty',
    text: 'No Results',
  }),
};

export default EmptyPage;
