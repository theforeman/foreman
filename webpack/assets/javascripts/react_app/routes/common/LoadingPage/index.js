import React from 'react';
import { Spinner } from 'patternfly-react';
import './loadingpage.scss';

const LoadingPage = () => (
  <div id="loading-page">
    <Spinner loading size="lg" />
  </div>
);

export default LoadingPage;
