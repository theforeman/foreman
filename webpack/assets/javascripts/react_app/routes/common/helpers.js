import { foremanUrl, urlBuilder } from '../../common/urlHelpers';

export const getControllerSearchProps = (
  controller,
  id = 'searchBar',
  canCreate = true
) => ({
  controller,
  autocomplete: {
    id,
    searchQuery: '',
    url: urlBuilder(controller, 'auto_complete_search'),
    useKeyShortcuts: true,
  },
  bookmarks: {
    url: foremanUrl('/api/bookmarks'),
    canCreate,
    documentationUrl: `4.1.5Searching`,
  },
});
