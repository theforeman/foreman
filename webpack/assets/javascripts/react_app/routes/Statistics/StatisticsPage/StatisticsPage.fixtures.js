import { statisticsMeta } from '../../../components/StatisticsChartsList/StatisticsChartsList.fixtures';
import { noop } from '../../../common/helpers';

export const discussionUrl =
  '/links/forums?post=t/trends-and-statistics-plugin/18745/4';

export const statisticsProps = {
  statisticsMeta,
  discussionUrl,
  isLoading: false,
  hasData: true,
  hasError: false,
  message: {},
  getStatisticsMeta: noop,
};
