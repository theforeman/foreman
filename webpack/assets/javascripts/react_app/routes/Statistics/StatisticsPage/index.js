import { compose, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { callOnMount } from '../../../common/HOC';

import * as actions from './StatisticsPageActions';
import withDataReducer from '../../common/reducerHOC/withDataReducer';
import {
  selectStatisticsMetadata,
  selectStatisticsMessage,
  selectStatisticsIsLoading,
  selectStatisticsHasMetadata,
  selectStatisticsHasError,
} from './StatisticsPageSelectors';

import StatisticsPage from './StatisticsPage';

// map state to props
const mapStateToProps = state => ({
  statisticsMeta: selectStatisticsMetadata(state),
  isLoading: selectStatisticsIsLoading(state),
  message: selectStatisticsMessage(state),
  hasData: selectStatisticsHasMetadata(state),
  hasError: selectStatisticsHasError(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { statisticsPage: withDataReducer('STATISTICS_PAGE') };

// export connected component
export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  callOnMount(({ getStatisticsMeta }) => getStatisticsMeta())
)(StatisticsPage);
