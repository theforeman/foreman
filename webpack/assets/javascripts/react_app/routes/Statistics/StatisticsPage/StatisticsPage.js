import React from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'patternfly-react';
import PageLayout from '../../common/PageLayout/PageLayout';
import Statistics from './Statistics/Statistics';
import { translate as __ } from '../../../common/I18n';

const StatisticsPage = ({ statisticsMeta, discussionUrl, ...props }) => (
  <PageLayout header={__('Statistics')} searchable={false}>
    <Alert type="warning">
      <span className="pficon pficon-warning-triangle-o" />
      <strong>This functionality is deprecated </strong>
      <span className="text">
        and will be removed in version 2.2. If you wish continue using it, you
        will need to install the Foreman Statistics plugin when upgrading to
        2.2.
        <a href={discussionUrl} target="_blank" rel="noreferrer">
          Join discussion
        </a>
      </span>
    </Alert>
    <Statistics statisticsMeta={statisticsMeta} {...props} />
  </PageLayout>
);

StatisticsPage.propTypes = {
  statisticsMeta: PropTypes.array,
  discussionUrl: PropTypes.string,
};

StatisticsPage.defaultProps = {
  statisticsMeta: [],
  discussionUrl: '',
};

export default StatisticsPage;
