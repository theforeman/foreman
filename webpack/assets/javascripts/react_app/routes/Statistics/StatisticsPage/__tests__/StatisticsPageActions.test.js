import API from '../../../../redux/API/API';

import { testActionSnapshotWithFixtures } from '../../../../common/testHelpers';
import { getStatisticsMeta } from '../StatisticsPageActions';
import { statisticsProps, discussionUrl } from '../StatisticsPage.fixtures';

jest.mock('../../../../redux/API/API');

const runStatisticsAction = (callback, props, serverMock) => {
  API.get.mockImplementation(serverMock);

  return callback(props);
};

const fixtures = {
  'should fetch statisticsMeta': () =>
    runStatisticsAction(getStatisticsMeta, {}, async () => ({
      data: {
        charts: statisticsProps.statisticsMeta,
        discussion_url: discussionUrl,
      },
    })),
  'should fetch statisticsMeta and fail': () =>
    runStatisticsAction(getStatisticsMeta, {}, async () => {
      throw new Error('some-error');
    }),
};

describe('StatisticsPage actions', () =>
  testActionSnapshotWithFixtures(fixtures));
