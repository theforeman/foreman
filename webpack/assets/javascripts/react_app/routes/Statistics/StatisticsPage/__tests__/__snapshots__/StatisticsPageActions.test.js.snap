// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPage actions should fetch statisticsMeta 1`] = `
Array [
  Array [
    Object {
      "type": "STATISTICS_PAGE_HIDE_LOADING",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "discussionUrl": "/links/forums?post=t/trends-and-statistics-plugin/18745/4",
        "hasData": true,
        "metadata": Array [
          Object {
            "id": "operatingsystem",
            "search": "/hosts?search=os_title=~VAL~",
            "title": "OS Distribution",
            "url": "statistics/operatingsystem",
          },
          Object {
            "id": "architecture",
            "search": "/hosts?search=facts.architecture=~VAL~",
            "title": "Architecture Distribution",
            "url": "statistics/architecture",
          },
        ],
      },
      "type": "STATISTICS_PAGE_DATA_RESOLVED",
    },
  ],
]
`;

exports[`StatisticsPage actions should fetch statisticsMeta and fail 1`] = `
Array [
  Array [
    Object {
      "type": "STATISTICS_PAGE_HIDE_LOADING",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "text": "some-error",
          "type": "error",
        },
      },
      "type": "STATISTICS_PAGE_FETCH_FAILED",
    },
  ],
]
`;
